site/docs/operations/gc-concerns.md

```diff
@@ -0,0 +1,108 @@
+---
+title: Concerns Related to GC
+description: Choosing the right JDK-GC combination is important.
+---
+
+With the introduction of the modern, fast-paced JDK release schedule the
+JVM landscape has undergone a phase change. In May 2020 we did an
+extensive test with various combinations of old and new JDK/GC
+combinations and we bring you the most important conclusions that should
+impact your Hazelcast Jet setup:
+
+1. As far as Hazelcast Jet is concerned, JDK version 8 is an antiquated
+   runtime and shouldn't be used. None of its garbage collectors have
+   satisfactory performance and their long GC pauses even result in
+   nodes being kicked out of the cluster.
+2. JDK 11 is the current Long-Term Support (LTS) version by Oracle and
+   it is the lowest version of JDK we recommend. We found that its
+   version of the G1, without any fine-tuning, already delivers
+   next-generation GC performance. The default maximum GC pause target
+   is 200 ms, which it meets in a very broad range of scenarios. By
+   configuring a single JVM option (`-XX:MaxGCPauseMillis`), you can
+   confine the maximum GC pause to 20-25 ms, which it can maintain at
+   lower GC pressure levels.
+4. There are two more GCs that are relevant, but as of this writing are
+   in the experimental phase. They are the low-latency collectors:
+   Oracle's Z and OpenJDK's Shenandoah. Testing on JDK 14, we found Z to
+   be the better of the two and a viable replacement for G1 in scenarios
+   that have low GC pressure, but very strict GC pause requirements. Z
+   was able to keep the maximum pause within 10 milliseconds. Since the
+   Z is available in OpenJDK as well, we recommend it for scenarios
+   where G1's GC pauses are too long. The version of Z in JDK 14 has
+   crude ergonomics to decide on a good background GC thread count, in
+   some cases you can improve the throughput by using
+   `-XX:ConcGCThreads`.
+
+Here are some more details, organized by the type of workload you are
+using Jet for.
+
+## Moderate State, Low-Latency Stream Processing
```

Comment:
Maybe we should quantify what's moderate or large state.

Reply:
I think the example below, 100,000 devices, quantifies it.

Reply:
It's much easier to quantify in GBs of heap data. I tried to calculate the state size for your example and there's a missing piece: the window size and lag. I guess it's a window of N seconds, sliding by 10-20ms and 100k keys. The 500Hz is irrelevant, it doesn't increase the state size. So if I assume 5-second lag and 5-second window and 32 bytes per entry, the formula is `100,000 * 1000/15 * 10 * 32`, so the state is 2GB, which is moderate. But if you use 1-minute smoothing window, the state is 13GB. And both answers are correct, because the window size isn't given. Maybe I also misestimated the per-element state: if you use string-encoded UUID for device ID, it will be much more...

What I think most people would do is to run jvisualvm, do a GC and see how much heap they need.

Reply:
GB is not a good quantification either because these costs scale with object count and not size. If you have GBs of large strings, it will slow down the GC about the same as MBs of small objects.

Reply:
Still it gives a better idea IMO. Ppl have no idea how much memory or how many objects are needed for "100,000 devices at 500Hz".